gadopt/momentum_equation.py

```diff
@@ -18,217 +18,175 @@
+            if compressible_stress:
+                trial_tensor_jump -= 2 / 3 * identity * (dot(eq.n, trial) - bc["un"])
+            trial_tensor_jump += transpose(trial_tensor_jump)
+            # Terms below are similar to the above terms for the DG dS integrals.
+            F += (
+                2
+                * sigma
+                * inner(outer(eq.n, eq.test), mu * trial_tensor_jump)
+                * eq.ds(bc_id)
+            )
+            F -= inner(mu * nabla_grad(eq.test), trial_tensor_jump) * eq.ds(bc_id)
+            # We only keep the normal part of stress; the tangential part is assumed to
+            # be zero stress (i.e. free slip) or prescribed via "stress".
+            F -= dot(eq.n, eq.test) * dot(eq.n, dot(stress, eq.n)) * eq.ds(bc_id)
+
+        if "stress" in bc:  # a momentum flux, a.k.a. "force"
+            # Here we need only the third term because we assume jump_u = 0
+            # (u_ext = trial) and stress = n . (mu . stress_tensor).
+            F -= dot(eq.test, bc["stress"]) * eq.ds(bc_id)
+
+        if "normal_stress" in bc:
+            F += dot(eq.test, bc["normal_stress"] * eq.n) * eq.ds(bc_id)
+
+        # if "drag" in bc:  # (bottom) drag of the form tau = -drag trial |trial|
+        #     drag = bc["drag"]
+        #     unorm = pow(dot(trial_lagged, trial_lagged) + 1e-6, 0.5)
+        #     F += dot(eq.test, drag * unorm * trial) * eq.ds(bc_id)
+
+    return -F
+
+
+def pressure_gradient_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    assert normal_is_continuous(eq.test)
+
+    F = -dot(div(eq.test), eq.p) * eq.dx
+
+    # Integration by parts gives a natural condition on pressure (as part of a normal
+    # stress condition). For boundaries where the normal component of u is specified, we
+    # remove that condition.
+    for bc_id, bc in eq.bcs.items():
+        if "u" in bc or "un" in bc:
+            F += dot(eq.test, eq.n) * eq.p * eq.ds(bc_id)
+
+    return -F
+
+
+def divergence_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    assert normal_is_continuous(eq.u)
+
+    rho = eq.approximation.rho
+    F = -dot(eq.test, div(rho * eq.u)) * eq.dx
+
+    # Add boundary integral for bcs that specify the normal component of u.
+    for bc_id, bc in eq.bcs.items():
+        if "u" in bc:
+            F -= eq.test * rho * dot(eq.n, bc["u"] - eq.u) * eq.ds(bc_id)
+        elif "un" in bc:
+            F -= eq.test * rho * (bc["un"] - dot(eq.n, eq.u)) * eq.ds(bc_id)
+
+    return -F
+
+
+def momentum_source_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
+    source = eq.approximation.buoyancy(eq.p, eq.T) * upward_normal(eq.mesh)
+    F = -dot(eq.test, source) * eq.dx
+
+    return -F
+
+
+residual_terms_momentum = [momentum_source_term, pressure_gradient_term, viscosity_term]
+residual_terms_stokes = [residual_terms_momentum, divergence_term]
```

Comment:
Should this be
```suggestion
residual_terms_stokes = residual_terms_momentum + [divergence_term]
```
to avoid nesting the lists of terms?

Reply:
Currently, nesting is desired because it separates the momentum and mass terms; I also allow terms to be either a list or a callable. It is true that we could also group all terms without changing the final UFL form. However, given that the mass equation should receive some additional terms in the near future, that may not be optimal.

Reply:
That's fair enough, if it's handled then I agree it gives a bit of semantic meaning to the list of terms.